### ISSUE TITLE
Out of memory in some tests due to memory limit confusion

### ISSUE THREAD

**Issue by Flamefire** [tensorflow/tensorflow#45664]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): CentOS 7
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): 2.4.0rc4
- Python version: 3.7.4
- Bazel version (if compiling from source): 3.4.1
- GCC/Compiler version (if compiling from source): GCC 8.3.0
- CUDA/cuDNN version: 10.1
- GPU model and memory: V100

**Describe the current behavior**

I have V100 GPUs with ~32GB memory. During startup of the test (many tests show this) I see lines like 
` W tensorflow/stream_executor/stream_executor_pimpl.cc:490] Not enough memory to allocate 31614597888 on device 0 within provided limit. [used=0, limit=1073741824]`

Some tests then fail after allocating about 1GB of memory trying to allocate more. The failure message includes the 31GB and shows almost 1GB as used.
E.g. //tensorflow/python/keras/applications:applications_test or //tensorflow/python/keras/layers:convolutional_recurrent_test

**Standalone code to reproduce the issue**
Run bazel test

**Other info / logs**
```
2020-12-14 12:27:32.863559: W tensorflow/stream_executor/stream_executor_pimpl.cc:490] Not enough memory to allocate 30637607680 on device 0 within provided limit. [used=0, limit=1073741824]
2020-12-14 12:27:32.863607: W tensorflow/stream_executor/stream_executor_pimpl.cc:490] Not enough memory to allocate 30637607680 on device 0 within provided limit. [used=0, limit=1073741824]
2020-12-14 12:27:42.864373: W tensorflow/stream_executor/stream_executor_pimpl.cc:490] Not enough memory to allocate 30637607680 on device 0 within provided limit. [used=0, limit=1073741824]
2020-12-14 12:27:42.864404: W tensorflow/stream_executor/stream_executor_pimpl.cc:490] Not enough memory to allocate 30637607680 on device 0 within provided limit. [used=0, limit=1073741824]
2020-12-14 12:27:42.864421: W tensorflow/core/common_runtime/bfc_allocator.cc:433] Allocator (GPU_0_bfc) ran out of memory trying to allocate 4.00MiB (rounded to 4194304)requested by op Tanh
Current allocation summary follows.
2020-12-14 12:27:42.864434: I tensorflow/core/common_runtime/bfc_allocator.cc:972] BFCAllocator dump for GPU_0_bfc
...
2020-12-14 12:27:42.866982: I tensorflow/core/common_runtime/bfc_allocator.cc:1040] Sum Total of in-use chunks: 928.00MiB
2020-12-14 12:27:42.866989: I tensorflow/core/common_runtime/bfc_allocator.cc:1042] total_region_allocated_bytes_: 976990208 memory_limit_: 31614597888 available bytes: 30637607680 curr_region_allocation_bytes_: 63229195776
2020-12-14 12:27:42.867000: I tensorflow/core/common_runtime/bfc_allocator.cc:1048] Stats: 
Limit:                     31614597888
InUse:                       973083648
MaxInUse:                    973083648
NumAllocs:                        2238
MaxAllocSize:                 86081536
Reserved:                            0
PeakReserved:                        0
LargestFreeBlock:                    0

2020-12-14 12:27:42.867013: W tensorflow/core/common_runtime/bfc_allocator.cc:441] ****************************************************************************************************
2020-12-14 12:27:42.867035: W tensorflow/core/framework/op_kernel.cc:1763] OP_REQUIRES failed at cwise_op_gpu_base.cc:97 : Resource exhausted: OOM when allocating tensor with shape[32,32,32,32] and type float on /job:localhost/replica:0/task:0/device:GPU:0 by allocator GPU_0_bfc

```

**Comment by amahendrakar** [comment]
@Flamefire,
In order to expedite the trouble-shooting process, could you please provide the exact sequence of commands / steps that you executed before running into the error. Thanks!

**Comment by Flamefire** [comment]
Sure, although the full command line is massive and I expect that to be a general error in the code or test, as the way TF was build shouldn't affect the limits it uses. Also upgrading to CUDA 11 is not possible on that machine yet, and also here I don't see how that would affect the limits.
Anyway full command line and config:
```
export CC_OPT_FLAGS='-O2 -ftree-vectorize -mcpu=native -fno-math-errno -fPIC'
export CUDA_TOOLKIT_PATH='/sw/installed/CUDA/10.1.243-GCC-8.3.0'
export CUDNN_INSTALL_PATH='/sw/installed/cuDNN/7.6.4.38-gcccuda-2019b'
export GCC_HOST_COMPILER_PATH='/scratch/ws/1/s3248973-EasyBuild/easybuild-ml/software/GCCcore/8.3.0/bin/gcc'
export GCC_HOST_COMPILER_PREFIX='/sw/installed/binutils/2.32-GCCcore-8.3.0/bin'
export MPI_HOME='/sw/installed/OpenMPI/3.1.4-gcccuda-2019b'
export NCCL_INSTALL_PATH='/sw/installed/NCCL/2.4.8-gcccuda-2019b'
export PYTHON_BIN_PATH='/scratch/ws/1/s3248973-EasyBuild/easybuild-ml/software/Python/3.7.4-GCCcore-8.3.0/bin/python'
export PYTHON_LIB_PATH='/tmp/ebinstall/software/TensorFlow/2.4.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages'
export TF_CUBLAS_VERSION='10.2.1'
export TF_CUDA_CLANG='0'
export TF_CUDA_COMPUTE_CAPABILITIES='3.7,6.1,7.0'
export TF_CUDA_PATHS='/sw/installed/CUDA/10.1.243-GCC-8.3.0'
export TF_CUDA_VERSION='10.1'
export TF_CUDNN_VERSION='7.6.4'
export TF_DOWNLOAD_CLANG='0'
export TF_ENABLE_XLA='1'
export TF_NCCL_VERSION='2.4.8'
export TF_NEED_AWS='0'
export TF_NEED_CUDA='1'
export TF_NEED_GCP='0'
export TF_NEED_GDR='0'
export TF_NEED_HDFS='0'
export TF_NEED_JEMALLOC='1'
export TF_NEED_KAFKA='0'
export TF_NEED_MPI='1'
export TF_NEED_OPENCL='0'
export TF_NEED_OPENCL_SYCL='0'
export TF_NEED_ROCM='0'
export TF_NEED_S3='0'
export TF_NEED_TENSORRT='0'
export TF_NEED_VERBS='0'
export TF_SET_ANDROID_WORKSPACE='0'
export TF_SYSTEM_LIBS='curl,boringssl,double_conversion,flatbuffers,gif,hwloc,icu,jsoncpp_git,libjpeg_turbo,png,lmdb,nasm,nsync,pcre,com_google_protobuf,pybind11,snappy,org_sqlite,zlib,absl_py,astor_archive,astunparse_archive,cython,dill_archive,enum34_archive,functools32_archive,gast_archive,opt_einsum_archive,pasta,six_archive,tblib_archive,termcolor_archive,typing_extensions_archive,wrapt'
export PYTHONPATH='/tmp/ebinstall/software/TensorFlow/2.4.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages:/tmp/ebinstall/software/TensorFlow/2.4.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages:/scratch/ws/1/s3248973-EasyBuild/easybuild-ml/software/typing-extensions/3.7.4.3-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages:/scratch/ws/1/s3248973-EasyBuild/easybuild-ml/software/flatbuffers-python/1.12-GCCcore-8.3.0-Python-3.7.4/lib/python3.7/site-packages:/scratch/ws/1/s3248973-EasyBuild/easybuild-ml/software/protobuf-python/3.10.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages:/sw/installed/h5py/2.10.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages:/sw/installed/SciPy-bundle/2019.10-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages:/scratch/ws/1/s3248973-EasyBuild/easybuild-ml/software/Python/3.7.4-GCCcore-8.3.0/easybuild/python'

bazel --output_user_root=/tmp/bazel-tf --host_jvm_args=-Xms512m --host_jvm_args=-Xmx4096m test --compilation_mode=opt --config=opt --subcommands --verbose_failures --config=noaws --jobs=64 --strip=never --copt="-ggdb" --copt="-fPIC" --action_env=CPATH='/sw/installed/cURL/7.66.0-GCCcore-8.3.0/include:/sw/installed/double-conversion/3.1.4-GCCcore-8.3.0/include:/scratch/ws/1/s3248973-EasyBuild/easybuild-ml/software/flatbuffers/1.12.0-GCCcore-8.3.0/include:/scratch/ws/1/s3248973-EasyBuild/easybuild-ml/software/giflib/5.2.1-GCCcore-8.3.0/include:/sw/installed/hwloc/1.11.12-GCCcore-8.3.0/include:/sw/installed/ICU/64.2-GCCcore-8.3.0/include:/scratch/ws/1/s3248973-EasyBuild/easybuild-ml/software/JsonCpp/1.9.3-GCCcore-8.3.0/include:/sw/installed/libjpeg-turbo/2.0.3-GCCcore-8.3.0/include:/sw/installed/libpng/1.6.37-GCCcore-8.3.0/include:/scratch/ws/1/s3248973-EasyBuild/easybuild-ml/software/LMDB/0.9.24-GCCcore-8.3.0/include:/scratch/ws/1/s3248973-EasyBuild/easybuild-ml/software/nsync/1.24.0-GCCcore-8.3.0/include:/sw/installed/PCRE/8.43-GCCcore-8.3.0/include:/sw/installed/protobuf/3.10.0-GCCcore-8.3.0/include:/sw/installed/pybind11/2.4.3-GCCcore-8.3.0-Python-3.7.4/include:/scratch/ws/1/s3248973-EasyBuild/easybuild-ml/software/snappy/1.1.7-GCCcore-8.3.0/include:/sw/installed/SQLite/3.29.0-GCCcore-8.3.0/include:/sw/installed/zlib/1.2.11-GCCcore-8.3.0/include' --action_env=LIBRARY_PATH='/sw/installed/cURL/7.66.0-GCCcore-8.3.0/lib:/sw/installed/double-conversion/3.1.4-GCCcore-8.3.0/lib:/scratch/ws/1/s3248973-EasyBuild/easybuild-ml/software/flatbuffers/1.12.0-GCCcore-8.3.0/lib64:/scratch/ws/1/s3248973-EasyBuild/easybuild-ml/software/giflib/5.2.1-GCCcore-8.3.0/lib:/sw/installed/hwloc/1.11.12-GCCcore-8.3.0/lib:/sw/installed/ICU/64.2-GCCcore-8.3.0/lib:/scratch/ws/1/s3248973-EasyBuild/easybuild-ml/software/JsonCpp/1.9.3-GCCcore-8.3.0/lib64:/sw/installed/libjpeg-turbo/2.0.3-GCCcore-8.3.0/lib64:/sw/installed/libpng/1.6.37-GCCcore-8.3.0/lib:/scratch/ws/1/s3248973-EasyBuild/easybuild-ml/software/LMDB/0.9.24-GCCcore-8.3.0/lib:/scratch/ws/1/s3248973-EasyBuild/easybuild-ml/software/nsync/1.24.0-GCCcore-8.3.0/lib64:/sw/installed/PCRE/8.43-GCCcore-8.3.0/lib:/sw/installed/protobuf/3.10.0-GCCcore-8.3.0/lib64:/scratch/ws/1/s3248973-EasyBuild/easybuild-ml/software/snappy/1.1.7-GCCcore-8.3.0/lib64:/sw/installed/SQLite/3.29.0-GCCcore-8.3.0/lib:/sw/installed/zlib/1.2.11-GCCcore-8.3.0/lib' --action_env=PYTHONPATH --action_env=PYTHONNOUSERSITE=1 --distinct_host_configuration=false --test_output=errors --test_env=TF_GPU_COUNT=6 --test_env=TF_TESTS_PER_GPU=1 --local_test_jobs=6 --build_tests_only --test_tag_filters='-no_gpu,-no_oss,-oss_serial,-benchmark-test,-no_oss_py37,-v1only' --build_tag_filters='-no_gpu,-no_oss,-oss_serial,-benchmark-test,-no_oss_py37,-v1only' --run_under=//tensorflow/tools/ci_build/gpu_build:parallel_gpu_execute -- //tensorflow/python/... -//tensorflow/python/integration_testing/...
```

This can of course be refined to only run 1 of the 2 mentioned tests which leads to the same issue.

**Comment by sanjoy** [comment]
Hi @Flamefire,

Please run the tests as follows:

```
bazel test \
  --build_tests_only \
\
  --test_env=TF_GPU_COUNT=8 \
  --test_env=TF_TESTS_PER_GPU=8 \
  --local_test_jobs=64 \
  --run_under=//tensorflow/tools/ci_build/gpu_build:parallel_gpu_execute \
\
  --nodistinct_host_configuration \
\
  --test_tag_filters=gpu,-no_gpu,-nogpu,-benchmark-test,-no_oss,-oss_serial,-v1only,-no_gpu_presubmit,-no_cuda11 \
\
  -- \
   //tensorflow/... \
  -//tensorflow/python/integration_testing/... \
  -//tensorflow/compiler/mlir/tosa/... \
  -//tensorflow/compiler/xrt/... \
  -//tensorflow/lite/micro/examples/... \
  -//tensorflow/core/tpu/... \
  -//tensorflow/lite/...
```

Where `TF_GPU_COUNT` is the number of GPUs available on your machine and the number passed to `--local_test_jobs` is `TF_GPU_COUNT * TF_TESTS_PER_GPU`.  Set `TF_TESTS_PER_GPU` low enough that your GPU has at least `2 * TF_TESTS_PER_GPU` GB of memory.

Of course the list of tests and the test tag filters should vary based on what you want to test.

**Comment by Flamefire** [comment]
I think I'm doing that already beeing even more conservative. The relevant parts from my above command also reformatted:

```
bazel \
--compilation_mode=opt --config=opt --config=noaws --jobs=64 \
--distinct_host_configuration=false \
--test_env=TF_GPU_COUNT=6 \
--test_env=TF_TESTS_PER_GPU=1 \
--local_test_jobs=6 \
--build_tests_only \
--test_tag_filters='-no_gpu,-no_oss,-oss_serial,-benchmark-test,-no_oss_py37,-v1only' \
--build_tag_filters='-no_gpu,-no_oss,-oss_serial,-benchmark-test,-no_oss_py37,-v1only' \
--run_under=//tensorflow/tools/ci_build/gpu_build:parallel_gpu_execute \
-- //tensorflow/python/... -//tensorflow/python/integration_testing/...
```

This is on a node with 6 GPUs and IIUC this runs 1 test per GPU only. So this is more conservative than your suggestion where I should be able to run 16 tests per GPU in parallel (32GB Volta)

Also quoting the message from the OP:
> W tensorflow/stream_executor/stream_executor_pimpl.cc:490] Not enough memory to allocate 31614597888 on device 0 within provided limit. [used=0, limit=1073741824]

This looks like only 1GB should be used at most for the tests due to that limit (although I'm unsure where this limit is set at)
So only explanation I have is that the test requires more than this 1GB

**Comment by sanjoy** [comment]
> I think I'm doing that already beeing even more conservative. The relevant parts from my above command also reformatted:

Thanks, the reformat made it much easier to read. :)

The full set of tags is `gpu,-no_gpu,-nogpu,-benchmark-test,-no_oss,-oss_serial,-v1only,-no_gpu_presubmit,-no_cuda11` (note that we only run tests that are explicitly tagged with `gpu`, amongst other constraints).  Other tests are not guaranteed to pass on GPUs.

And you'll notice that e.g. `//tensorflow/python/keras/applications:applications_test` is not tagged with `gpu`.  So the command line I proposed above will filter that test out.

**Comment by Flamefire** [comment]
> Other tests are not guaranteed to pass on GPUs.

This makes me very surprised. My understanding of unit tests is, that they should test code for correctness. The availability of GPUs is optional in "normal" usage of TensorFlow.
How can a test fail when a GPU is available? Doesn't that mean that either the TF code it tests has a bug (i.e. the test code uses it like a user would but it returns wrong results) or that the test code has a bug (e.g. it runs TF code that is only there for CPU and the test fails to set CPU as the device)?

To put it another way: My expectation would be: When running unit tests, all tests are run by default. When some tests require additional constraints (e.g. a GPU or a certain amount of memory) they are marked as such and automatically skipped when it is detected that the current system doesn't has this. The filter-tags allow to limit the amount of tests gathered to exclude e.g. long-running tests (e.g. benchmarks) or to not consider the GPU tests at all (as you know already they will be skipped)

My use case is that I want to run the TF unit tests at the end of a build on HPC clusters. The whole install process is heavily automated as it is unfeasible for HPC admins worldwide to manually go through each softwares unique build process. Here it is not known if the build machine contains GPUs or not. Hence I would like to additionally run the GPU tests if there are GPUs available. If now some tests fail only because a GPU is available then it renders the units tests unusable in that scenario and IMO the tests should be fixed.

**Comment by sanjoy** [comment]
> How can a test fail when a GPU is available?

In this specific case it looks like the test needs more than 1GB to run.

> Hence I would like to additionally run the GPU tests if there are GPUs available. If now some tests fail only because a GPU is available then it renders the units tests unusable in that scenario and IMO the tests should be fixed.

I don't think we'll be able to change how we run our tests anytime soon.  (We could make spot fixes here and there, but these kinds of things will likely regress over time.)

So I recommend you do two runs, one without GPUs and one with GPUs.  The first one excludes all tests tagged with `gpu` and run with `CUDA_VISIBLE_DEVICES=-1` to ensure that the tests don't see any GPUs.  The second one runs only GPU tests, and runs only if GPUs are present on the machine.

**Comment by Flamefire** [comment]
> So I recommend you do two runs, one without GPUs and one with GPUs

Although inconvenient (and I wonder what the `nogpu` and `no_gpu` tags are for then) I tried that. So based on the CI files I have 
```
        'test_tag_filters_gpu': 'gpu,-no_gpu,-nogpu,-no_oss,-oss_serial,-benchmark-test,-v1only',
        'test_tag_filters_cpu': '-gpu,-tpu,-no_oss,-oss_serial,-benchmark-test,-v1only',
```

However there are failures in the CPU tests now. E.g. `//tensorflow/core/common_runtime:ring_reducer_test` tries to run `RingReducerTest.DaTyFLOAT_DevTyGPU_Wkr1_Dev2_Sdiv1_Len1_Abrt0` which then fails with "F tensorflow/core/common_runtime/ring_reducer_test.cc:187] Unsupported device_type GPU"

From checking the BUILD files I see that there is a ring_reducer_test and ring_reducer_test_gpu added through tf_cuda_cc_test

And in the cc file the GPU section is guarded by `#if GOOGLE_CUDA`, but as I verified that ring_reducer_test (w/o _gpu suffix) is run there seems to be GOOGLE_CUDA wrongly(?) defined.
Can you advise what to do now? Why and where is GOOGLE_CUDA defined? And why am I seeing a GPU related failure in a supposedly CPU test which seems to not happen on the TF CI? Is that test excluded there?

**Comment by Flamefire** [comment]
Ok this seems to be another issue which can be workaround by excluding , see

In a similar case the presence of the `no_gpu` (or `nogpu`, why 2?) is odd. As noted in https://github.com/tensorflow/tensorflow/issues/45664#issuecomment-755798751 it seems to be unsupported or impossible to run tests *not* marked with `gpu` in an environment which has GPUs. Hence tests need to be started with a test_tag_filter of either `gpu` or `-gpu` anyway, so what are the `no_gpu` tags for?
May I suggest to remove those? Or could you clarify the meaning of those?

**Comment by sanjoy** [comment]
The test tags could use some cleaning up, but that's more complicated than it sounds since there are several bazel macros involved here (and more blaze macros internally at Google).  So I suggest not blocking progress on this.

Regarding `GOOGLE_CUDA` -- it is set via `--config=cuda`.  The non-GPU tests should not be built or tested with `--config=cuda`.  The `configure.py` workflow makes this difficult by writing out `build --config=cuda` to `.tf_configure.bazelrc` the user asked for it.  This means so if you configure a source tree on a machine with a GPU present (and enable CUDA) and build/test on a machine without a GPU then `--config=cuda` is set even for the second run.  Building & running TensorFlow itself this way is fine (TF built with `--config=cuda` silently falls back to the CPU if no GPU is present), but unit tests may fail when run this way.  So I would recommend reconfiguring TF before running the non-GPU test suite.

**Comment by Flamefire** [comment]
>  So I would recommend reconfiguring TF before running the non-GPU test suite.

This is highly inconvenient. Doing it this way I can't be sure that the TF I built (and installed) is the one that is tested as (I think) if I reconfigure then Bazel will rebuild also the core TF parts which are being tested.
IMO the correct approach would be to not rely on a flag/define which is intended to signal GPU support (as you wrote it is fine to have GPU support but no GPU) in a test which is meant to test the CPU implementation (on which TF should fall back). This basically makes it impossible, to test if the CPU fallback is (also) correctly built.

So I would suggest to add a new flag which tells the tests that they are CPU tests (I mean: those are even added twice: Once for cpu once for gpu and a workaround via a test tag was neccessary to "fix" this), see my comment in #46425

**Comment by mihaimaruseac** [comment]
Somehow this slipped. I think with the upcoming refactor we can target this too

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.

**Comment by Flamefire** [comment]
@mihaimaruseac Has there been any progress in the refactoring you mentioned?

**Comment by mihaimaruseac** [comment]
I left the team shortly after that mention and I lost track. I'm only helping here and there based on what I still know is happening, but I no longer have full visibility
